### PR TITLE
Storage: Refactor DMFileReader

### LINKED ISSUE #8904: TiFlash result is wrong
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

prepare some data
do random dml and check `select count(*)` using tiflash and tikv
upgrade from 6.17 to master, when one tiflash version is 6.1.7 and another is master, query from tiflash is wrong
```
mysql> begin;
Query OK, 0 rows affected (0.04 sec)

mysql> set tidb_isolation_read_engines='tiflash';
Query OK, 0 rows affected (0.03 sec)

mysql> select count(*) from t;
+----------+
| count(*) |
+----------+
| 15284754 |
+----------+
1 row in set (0.17 sec)

mysql> set tidb_isolation_read_engines='tikv';
Query OK, 0 rows affected (0.03 sec)

mysql> select count(*) from t;
+----------+
| count(*) |
+----------+
| 16117328 |
+----------+
1 row in set (0.15 sec)

mysql> commit;
Query OK, 0 rows affected (0.04 sec)
```

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: close #8904

Problem Summary:

In the previous implementation, I used `insertManyFrom` to insert data in ConstColumn into the target column.

https://github.com/pingcap/tiflash/blob/adc57e22149d04d76d26f5193134a05a938af7c7/dbms/src/Columns/ColumnVector.h#L216-L220

In `insertManyFrom`, ColumnVector::insertManyFrom only expected calling src with exactly the same type of Self rather than any subclass of IColumn. If we do so, static_cast<const Self &>(src).getData() just returns a random memory location. It causes the del tag column contains too much non-zero value, leading to return less data than expected.

### What is changed and how it works?

- Add a new interface `insertMany`  in `IColumn` https://github.com/pingcap/tiflash/pull/8925
- Use `insertMany` instead of `insertManyFrom`

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by CalvinNeo** [comment]
Could you elaborate the bug in this issue

**Comment by Lloyd-Pottiger** [comment]
> Could you elaborate the bug in this issue

done

**Comment by JaySon-Huang** [comment]
> In insertManyFrom, the data is get from the container, but ConstColumn does not have any container.

Actually, `ColumnVector::insertManyFrom` only expected calling `src` with exactly the same type of `Self` rather than any subclass of `IColumn`. If we do so, `static_cast<const Self &>(src).getData()` just returns a random memory location.

**Comment by Lloyd-Pottiger** [comment]
> > In insertManyFrom, the data is get from the container, but ConstColumn does not have any container.
> 
> Actually, `ColumnVector::insertManyFrom` only expected calling `src` with exactly the same type of `Self` rather than any subclass of `IColumn`. If we do so, `static_cast<const Self &>(src).getData()` just returns a random memory location.

updated

**Comment by JaySon-Huang** [comment]
/hold
need to verify with coverage report

**Comment by Lloyd-Pottiger** [comment]
image

image

File: release-centos7-llvm/gen_coverage.sh [hunk 1]
```diff
@@ -54,7 +54,7 @@ cd -
 ## TODO: Add more testing binaries
 
 # Collect the prof raw data and generate cov report
-llvm-profdata merge -sparse "${COVERAGE_DIR}/*.profraw" -o "${COVERAGE_DIR}/merged.profdata"
+llvm-profdata merge -sparse ${COVERAGE_DIR}/*.profraw -o "${COVERAGE_DIR}/merged.profdata"
```
Comment (JaySon-Huang) [inline release-centos7-llvm/gen_coverage.sh:57]: why the `""` is removed?

File: dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp [hunk 2]
```diff
@@ -142,156 +117,158 @@ bool DMFileReader::getSkippedRows(size_t & skip_rows)
...
+            else
+            {
+                for (size_t i = 0; i < block.columns(); ++i)
+                {
+                    columns[i]->insertRangeFrom(
+                        *block.getByPosition(i).column,
+                        0,
+                        block.getByPosition(i).column->size());
+                }
+            }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp:253]: We assume `readWithFilter` will never be called on the three hidden columns, so we don't need to take care about whether the `block.getByPosition(i).column` is ColumnConst or not?

File: dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp [hunk 3]
```diff
@@ -142,156 +117,158 @@ bool DMFileReader::getSkippedRows(size_t & skip_rows)
...
-            blocks.emplace_back(std::move(block));
+            else
+            {
+                for (size_t i = 0; i < block.columns(); ++i)
+                {
+                    columns[i]->insertRangeFrom(
+                        *block.getByPosition(i).column,
+                        0,
+                        block.getByPosition(i).column->size());
+                }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp:252]: Are there any cases cover this branch?

File: release-centos7-llvm/gen_coverage.sh (hunk 1 above)
Comment (Lloyd-Pottiger) [reply release-centos7-llvm/gen_coverage.sh:57]: add `""` will not be viewed as regrex, cause "{COVERAGE_DIR}/*.profraw no such file or directory"

File: dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp (hunk 2 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp:253]: image

clean read is not enabled when there are hidden columns.

File: dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp (hunk 3 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp:252]: It is kind of hard to construct.... https://github.com/pingcap/tiflash/pull/8927#discussion_r1566627981 never do clean read here, so the source and target column are must be the same.

File: dbms/src/Storages/DeltaMerge/File/ColumnCache.cpp [hunk 4]
```diff
@@ -55,8 +54,53 @@ RangeWithStrategys ColumnCache::getReadStrategy(size_t pack_id, size_t pack_coun
...
+    auto strategy = Strategy::Unknown;
+    size_t range_start = 0;
+    for (size_t cursor = target_range.first; cursor < target_range.second; ++cursor)
+    {
+        if (std::find(clean_read_pack_idx.cbegin(), clean_read_pack_idx.cend(), cursor) != clean_read_pack_idx.cend())
+        {
+            if (strategy == Strategy::Memory)
+            {
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/File/ColumnCache.cpp:80]: image

done

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**